appliance/mariadb/process.go

```diff
@@ -0,0 +1,1106 @@
+				return err
+			}
+
+			// TODO(benbjohnson): Unlock tables after backup and log pos are retrieved.
+
+			return nil
+		}(); err != nil {
+			if files, err := ioutil.ReadDir("/data"); err == nil {
+				for _, file := range files {
+					os.RemoveAll(filepath.Join("/data", file.Name()))
+				}
+			}
+			return err
+		}
+	}
+
+	if err := p.start(); err != nil {
+		return err
+	}
+
+	if err := func() error {
+		logger.Info("retrieving upstream log coordinates")
+		ms, err := p.masterStatus(upstream.Addr)
+		if err != nil {
+			logger.Error("error reading upstream log coordinates", "err", err)
+			return err
+		}
+
+		// Connect to local server and set up slave replication.
+		dsn := p.DSN("127.0.0.1:"+p.Port, "")
+		dsn.User = "root"
+		dsn.Password = ""
+
+		db, err := sql.Open("mysql", dsn.String())
+		if err != nil {
+			logger.Error("error acquiring connection", "err", err)
+			return err
+		}
+		defer db.Close()
+
+		host, port, _ := net.SplitHostPort(upstream.Addr)
+		logger.Info("changing master", "host", host, "port", port)
+		if _, err := db.Exec(fmt.Sprintf(
+			`CHANGE MASTER TO MASTER_HOST='%s', MASTER_PORT=%s, MASTER_USER='flynn', MASTER_PASSWORD='%s', MASTER_LOG_FILE='%s', MASTER_LOG_POS=%d, MASTER_CONNECT_RETRY=10;`,
+			host,
+			port,
+			p.Password,
+			ms.File,
+			ms.Position,
+		)); err != nil {
+			logger.Error("error changing master", "host", host, "port", port, "file", ms.File, "position", ms.Position, "err", err)
+			return err
+		}
+
+		// Start slave.
+		logger.Info("starting slave")
+		if _, err := db.Exec(`START SLAVE`); err != nil {
+			return err
+
+		}
+
+		// Install semi-sync slave plugin.
+		if _, err := db.Exec(`INSTALL PLUGIN rpl_semi_sync_slave SONAME 'semisync_slave.so'`); err != nil {
+			logger.Error("error installing rpl_semi_sync_slave", "err", err)
+			return err
+		}
+		if _, err := db.Exec(`STOP SLAVE IO_THREAD`); err != nil {
+			logger.Error("error stopping slave io thread", "err", err)
+			return err
+		}
+		if _, err := db.Exec(`START SLAVE IO_THREAD`); err != nil {
+			logger.Error("error starting slave io thread", "err", err)
+			return err
+		}
+
+		// Enable semi-synchronous on slave.
+		if _, err := db.Exec(`SET GLOBAL rpl_semi_sync_slave_enabled = ON`); err != nil {
+			return err
+		}
+
+		return nil
+	}(); err != nil {
+		return err
+	}
+
+	if downstream != nil {
+		p.waitForSync(downstream, false)
+	}
+
+	return nil
+}
+
+// initPrimaryDB initializes the local database with the correct users and plugins.
+func (p *Process) initPrimaryDB() error {
+	logger := p.Logger.New("fn", "initPrimaryDB")
+	logger.Info("initializing primary database")
+
+	dsn := p.DSN("127.0.0.1:"+p.Port, "")
+	dsn.User = "root"
+	dsn.Password = ""
+
+	db, err := sql.Open("mysql", dsn.String())
+	if err != nil {
+		logger.Error("error acquiring connection", "err", err)
+		return err
+	}
+	defer db.Close()
+
+	if _, err := db.Exec(fmt.Sprintf(`CREATE USER 'flynn'@'%%' IDENTIFIED BY '%s'`, p.Password)); err != nil {
```

Comment:
I still need to handle the error. I was working on getting the happy path working first.